### PR TITLE
fetch: add data: & about: uri support

### LINKED ISSUE #1164: Support `createObjectURL` based `blob:` URLs in node 16+
Node's experimental [URL.createObjectUrl](https://nodejs.org/api/url.html#urlcreateobjecturlblob) creates URLs to blob data that look something like this: `blob:nodedata:ee7f0009-a19f-47f3-935f-08fa7ca35adc`

In the browser environment `fetch` and friends will accept `blob:` URLs to allow applications to pass URL references to large Blobs of data without needing to pull them into data URIs.

For me this is especially useful when working with libraries like THREE.js which make heavy use of fetch APIs - which then need to be monkey-patched in NodeJS

## Logs & Screenshots

What I'm seeing atm:

```
(node:18940) ExperimentalWarning: --experimental-loader is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
(node:18940) DeprecationWarning: Obsolete loader hook(s) supplied and will be ignored: getFormat, transformSource
Promise { <pending> }
/home/wstott/repos/tritium/libs/typescript/tritium_3d_view/node_modules/undici/lib/fetch/index.js:176
        Object.assign(new TypeError('fetch failed'), { cause: response.error })
                      ^

TypeError: fetch failed
    at Object.processResponse (/home/wstott/repos/tritium/libs/typescript/tritium_3d_view/node_modules/undici/lib/fetch/index.js:176:23)
    at fetchFinale (/home/wstott/repos/tritium/libs/typescript/tritium_3d_view/node_modules/undici/lib/fetch/index.js:737:17)
    at Fetch.mainFetch (/home/wstott/repos/tritium/libs/typescript/tritium_3d_view/node_modules/undici/lib/fetch/index.js:712:5) {
  cause: Error: URL scheme must be a HTTP(S) scheme
      at makeNetworkError (/home/wstott/repos/tritium/libs/typescript/tritium_3d_view/node_modules/undici/lib/fetch/response.js:364:11)
      at /home/wstott/repos/tritium/libs/typescript/tritium_3d_view/node_modules/undici/lib/fetch/index.js:560:16
      at Fetch.mainFetch (/home/wstott/repos/tritium/libs/typescript/tritium_3d_view/node_modules/undici/lib/fetch/index.js:583:7)
      at Fetch.fetching (/home/wstott/repos/tritium/libs/typescript/tritium_3d_view/node_modules/undici/lib/fetch/index.js:452:20)
      at Agent.fetch (/home/wstott/repos/tritium/libs/typescript/tritium_3d_view/node_modules/undici/lib/fetch/index.js:195:6)
      at fetch (/home/wstott/repos/tritium/libs/typescript/tritium_3d_view/node_modules/undici/index.js:102:22)
      at FileLoader.load (/home/wstott/repos/tritium/libs/typescript/tritium_3d_view/node_modules/three/build/three.js:28384:4)
      at GLTFLoader.load (file:///home/wstott/repos/tritium/libs/typescript/tritium_3d_view/node_modules/three/examples/jsm/loaders/GLTFLoader.js:190:10)
      at file:///home/wstott/repos/tritium/libs/typescript/tritium_3d_view/tools/gltf2sdf.ts:26:14
      at new Promise (<anonymous>)
}

```

I'm assuming by  `URL scheme must be a HTTP(S) scheme` blob schemes are simply not implemented, as opposed to there being a bug.

## Environment

Ubuntu 18.04 LTS, Node v17.3.1, Undici 4.12.1

### PR DESCRIPTION
**Pull request by KhafraDev**
Closes #1164

### CODE REVIEW

**Comment by KhafraDev** [comment]
It would appear as though I wasn't the only one who had issues with the spec in this case https://github.com/nodejs/undici/issues/929#issuecomment-895745618

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by KhafraDev** [comment]
Tried looking into this again and this is as far as I got (spec dump coming up):

- [A local scheme](https://fetch.spec.whatwg.org/#local-scheme) is one with its origin as `about:`, `blank:`, or `data:`. Likewise, a local url is a URL with a local scheme.
- The term "local scheme" is never mentioned again in the spec.
- [request.localURLsOnly](https://fetch.spec.whatwg.org/#local-urls-only-flag) is a flag that **should** be set when a URL is a local url.
- `localURLsOnly` is mentioned [one](https://fetch.spec.whatwg.org/#main-fetch) time in the spec and is never set. Undici [handles](https://github.com/nodejs/undici/blob/main/lib/fetch/index.js#L477) this case already, although it will never be true.
- Then, I tried looking at the URL spec (the "URLs" referred to in the spec are NOT whatwg URLs), but it also did not mention anything about local URLs.

So to sum it up, no progress has been made.

**Comment by KhafraDev** [comment]
Fixed two pretty minor bugs with this dive at least 🗡️ 

- Undici was incorrectly [comparing origins](https://html.spec.whatwg.org/multipage/origin.html#same-origin) in multiple ways. The whatwg URL origin is different than the fetch URL origin (it would be the whatwg URL's host[name]) and it wouldn't perform all 3 checks required.
- `response` was never set when it should have and the function was returning before all steps were performed.

**Comment by ronag** [comment]
> These changes should probably land regardless of this PR?

Please make a separate PR for those.

**Comment by KhafraDev** [comment]
Added the remaining WPT tests (surprisingly everything other than 1 test already worked without changes) but I haven't had any luck figuring out where the spec calls scheme fetch when fetch's `init` is undefined!

On the other hand, something like
```js
const res = await fetch(blobURL, { mode: 'no-cors' })
```
will work without my hack, but this is not in line with FireFox/Chromium.

**Comment by ronag** [comment]
Good job!
